File: ALGORITHM/Leetcode/java/001_Two_Sum.java
/**
 * @FileName : 001_Two_Sum.py
 * @Project : LeetCode
 * @Date : 2020-08-27
 * @author : User

 * @Description :
 Approach 1 : Brute Force
 Approach 2 : Two-pass Hash Table
 Approach 3 : One-pass Hash Table
 */

class Solution {
    public static void main(String[] args) {
        Solution s = new Solution();

        s.twoSumBruteForce();
        s.twoSum2PassHash();
        s.twoSum();
    }

     // Approach 1
     // Time complexity : O(n^2)
     // Space complexity = O(1)
    public int[] twoSumBruteForce(int[] nums, int target) {
        for(int i=0; i<nums.length; ++i) {
            for(int j=i+1; j<nums.length; ++j) {
                if(nums[i] + nums[j] == target)
                    return new int[] {i, j};
            }
        }
    }

     // Approach 2
     // Time complexity : O(n)
     // Space complexity = O(n)
    public int[] twoSum2PassHash(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();

     }


     // Approach 3
     // Time complexity : O(n^2)
     // Space complexity = O(1)
    public int[] twoSum(int[] nums, int target) {
        for(int i=0; i<nums.length; ++i) {
            for(int j=i+1; j<nums.length; ++j) {
                if(nums[i] + nums[j] == target)
                    return new int[] {i, j};
            }
        }
    }
}

File: Leetcode/python/001_Two_Sum.py
"""
 * @FileName : 001_Two_Sum.py
 * @Project : LeetCode
 * @Date : 2020-08-27
 * @author : User

 * @Description : java 파일에서 3가지 방법을 모두 보였음으로 최적의 방법만 메모
 * --> One-pass Hash Table

"""

from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        sum_dict = {}

        for i, num in enumerate(nums):
            complement = target - num

            if complement in sum_dict:
                return sum_dict[complement], i
            sum_dict[num] = i

